### PR TITLE
spacy: added Explosion AI spaCy, thinc and related tools 

### PR DESCRIPTION
**Pull request by aptlin**
###### Motivation for this change

Making NLP analysis easier with Nix.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/python-modules/cymem.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  
+  checkPhase = ''
+    cd cymem/tests
+    ${python.interpreter} -m unittest discover -p "*test*"
+  '';
+  
+  meta = with stdenv.lib; {
+    description = "Cython memory pool for RAII-style memory management";
+    homepage = https://github.com/explosion/cymem;
+    platforms = platforms.unix;
```
Comment (FRidh) [inline pkgs/development/python-modules/cymem.nix]: don't set platforms in Python packages. The function already takes care of that.

File: pkgs/development/python-modules/ftfy.nix [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  src = fetchurl {
+    url = "mirror://pypi/f/ftfy/${name}.tar.gz";
+    sha256 = "0pqnzgipfjb92fcyqklkg8zqfm1w596rngw31q5hq0qdap3w70hd";        
+  };
+
+  propagatedBuildInputs = [ html5lib wcwidth];
+
+  buildInputs = [
+    pytest
+    nose
```
Comment (FRidh) [inline pkgs/development/python-modules/ftfy.nix]: It needs pytest and nose, but uses unittest as runner? That's odd.

File: pkgs/development/python-modules/ftfy.nix [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+, wcwidth
+, nose
+, pytest
+, python
+}:
+buildPythonPackage rec {
+  name = "ftfy-${version}";
+  version = "4.4.2";
+
+  src = fetchurl {
```
Comment (FRidh) [inline pkgs/development/python-modules/ftfy.nix]: fetchPypi

File: pkgs/development/python-modules/murmurhash.nix [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  
+  src = fetchFromGitHub {
+    owner = "explosion";
+    repo = "murmurhash";
+    rev = "0.26.4";
+    sha256 = "0n2j0glhlv2yh3fjgbg4d79j1c1fpchgjd4vnpw908l9mzchhmdv";    
+  };
+
+  propagatedBuildInputs = [
+   cython
```
Comment (FRidh) [inline pkgs/development/python-modules/murmurhash.nix]: This is not a runtime dependency, or is it?

**Review by FRidh — CHANGES_REQUESTED** [review]
- Looks good overall
- Remove meta.platforms
- Preferably use fetchPypi
- Do not commit whitespace changes in `python-packages.nix`

**Comment by sdll** [comment]
@FRidh, thank you very much for your review!

 I have made requested changes and cleaned up the code, removing whitespace trimming in python-packages.nix.

 With regards to murmurhash, cython is a build dependency, as you can see [here](https://github.com/explosion/murmurhash/blob/master/requirements.txt).

**Comment by FRidh** [comment]
> With regards to murmurhash, cython is a build dependency, as you can see here.

A requirements.txt file doesn't say much, because its not used when building a package with setuptools, unless its explicitly read and added to e.g.  `install_requires`. Looking at the `setup.py` file it seems it is only required to run setup and perform the build. 

If a package is only used during build and install time, then it should be added to `buildInputs`. If it is needed at runtime or when another package that depends on it is build, it should be added to `propagatedBuildInputs`.

**Comment by sdll** [comment]
@FRidh, thank you for teaching me something new today!

I have changed ftfy to make Travis CI builds work for GNU/Linux and OS X.

 I have also changed murmurhash build inputs as requested.

**Comment by sdll** [comment]
@FRidh, I moved recipes to satisfy the 'package/default.nix' format, as well as replaced `with rec {....}` with `let ... in`. 

OS X seems to time out, could you please tell me how I can decrease the build time? My first idea was to set `doCheck = false;` in some recipes exclusively for OS X, is there a way to do it?

**Comment by FRidh** [comment]
@sdll you can ignore Travis. While it is sometimes useful, we cannot really rely on it.

**Comment by sdll** [comment]
@FRidh, is there anything else I need to change? Is the tag "needs: clean-up" still relevant?

**Comment by sdll** [comment]
@FRidh, I have updated the recipe ftfy, the testing issue was fixed (see https://github.com/LuminosoInsight/python-ftfy/issues/75)

**Comment by sdll** [comment]
@FRidh, 

Travis CI on OS X seems to fail again due to the same error as in https://github.com/NixOS/nixpkgs/pull/25804:
```
pyemd/emd.cpp:458:10: fatal error: 'vector' file not found
```
Similarly, do you think we should try to fix it by specifying GCC_INSTALL_HOME?
See a similar question discussed here: ysshao/LLVM-Tracer#7

Is there a deeper problem with how clang c++ headers are treated in the OS X build of Nix?

**Comment by FRidh** [comment]
@knedlsepp is this related to #27649? (I'm just diving through old PR's)

**Comment by knedlsepp** [comment]
@FRidh: Seems highly likely. I'll try to rework my PR asap, so we can merge.

**Comment by FRidh** [comment]
This needs a minor rebase. I successfully built the 2.7 version of spacy but the 3.6 hang during building.

**Comment by FRidh** [comment]
I've pushed this to master.

**Comment by JonathanReeve** [comment]
Great work! Any idea how to install the data using Nix? (E.g. `en_core_web_lg`?)

**Pull request closed without merging**